packages/flutter_tools/lib/src/protocol_discovery.dart

```diff
@@ -49,50 +57,70 @@ class ProtocolDiscovery {
   final int devicePort;
   final bool ipv6;
 
-  final Completer<Uri> _completer = Completer<Uri>();
+  /// The time to wait before forwarding a new observatory URIs from [logReader].
+  final Duration throttleDuration;
 
   StreamSubscription<String> _deviceLogSubscription;
+  StreamController<Uri> _uriStreamController;
 
   /// The discovered service URI.
+  /// Use [uris] instead.
+  // TODO(egarciad): replace `uri` for `uris`.
+  Future<Uri> get uri {
+    return uris.first;
+  }
+
+  /// The discovered service URI stream.
   ///
   /// Port forwarding is only attempted when this is invoked, in case we never
   /// need to port forward.
-  Future<Uri> get uri async {
-    final Uri rawUri = await _completer.future;
-    return await _forwardPort(rawUri);
+  ///
+  /// Dependending on the lifespan of the app running the observatory,
+  /// a new observatory URI may be assigned to the app.
+  Stream<Uri> get uris {
+    return _uriStreamController.stream
+      .transform(_throttle<Uri>(
+        waitDuration: throttleDuration,
+      ))
+      .asyncMap<Uri>((Uri observatoryUri) => _forwardPort(observatoryUri));
```

Comment:
Further golfing using a tear-off:

```suggestion
      .asyncMap<Uri>(_forwardPort);
```